docs/source/training.md

```diff
@@ -0,0 +1,196 @@
+
+### Example `launch.sh` Script for Derecho
+
+```bash
+#!/bin/bash
+#PBS -A NAML0001
+#PBS -N train_model
+#PBS -l walltime=12:00:00
+#PBS -l select=1:ncpus=64:ngpus=4
+#PBS -q main
+#PBS -j oe
+#PBS -k eod
+#PBS -r n
+
+# Load modules
+module purge
+module load ncarenv/23.09
+module reset
+module load gcc craype cray-mpich cuda cudnn/8.8.1.3-12 conda
+conda activate credit-derecho
+
+# Export environment variables
+export LSCRATCH=/glade/derecho/scratch/schreck/
+export LOGLEVEL=INFO
+export NCCL_DEBUG=INFO
+export CUDA_VISIBLE_DEVICES=0,1,2,3
+export NCCL_SOCKET_IFNAME=hsn
+export MPICH_GPU_MANAGED_MEMORY_SUPPORT_ENABLED=1
+export MPICH_OFI_NIC_POLICY=GPU
+export MPICH_GPU_SUPPORT_ENABLED=1
+export NCCL_IB_DISABLE=1
+export NCCL_CROSS_NIC=1
+export NCCL_NCHANNELS_PER_NET_PEER=4
+export MPICH_RDMA_ENABLED_CUDA=1
+export NCCL_NET="AWS Libfabric"
+export NCCL_NET_GDR_LEVEL=PBH
+export FI_CXI_DISABLE_HOST_REGISTER=1
+export FI_CXI_OPTIMIZED_MRS=false
+export FI_MR_CACHE_MONITOR=userfaultfd
+export FI_CXI_DEFAULT_CQ_SIZE=131072
+
+# Log GPU and node details
+echo "Number of nodes: 1"
+echo "Number of GPUs per node: 32"
+echo "Total number of GPUs: 4"
+
+# Find head node's IP
+nodes=( $(cat $PBS_NODEFILE) )
+head_node=${nodes[0]}
+head_node_ip=$(ssh $head_node hostname -i | awk '{print $1}')
+
+# Launch training
+MASTER_ADDR=$head_node_ip MASTER_PORT=1234 mpiexec -n 4 --ppn 4 --cpu-bind none \
+    python applications/train.py -c model.yml --backend nccl
+```
+
+This script utilizes **MPI** for coordinating training across **multiple nodes and GPUs**. It includes necessary environment variables for Derecho’s system configuration. **Users should not need to modify this script**, as it is tailored for Derecho and may change with system updates.
+
+## Running on Casper vs. Derecho
+
+For **Casper**, modify `model.yml` as follows:
+
+```yaml
+pbs:
+    conda: "credit"
+    project: "NAML0001"
+    job_name: "train_model"
+    nodes: 1
+    ncpus: 8
+    ngpus: 1
+    mem: '128GB'
+    walltime: '4:00:00'
+    gpu_type: 'a100'
+    queue: 'casper'
+```
+
+Once again, to launch the job on Casper, run:
+
+```bash
+python applications/train.py -c config/model.yml -l 1
+```
+
+This command generates a **launch script (`launch.sh`)**, which will look like:
+
+```bash
+#!/bin/bash -l
+#PBS -N train_model
+#PBS -l select=1:ncpus=8:ngpus=1:mem=128g
+#PBS -l walltime=4:00:00
+#PBS -l gpu_type=a100
+#PBS -A NAML0001
+#PBS -q casper
+#PBS -j oe
+#PBS -k eod
+source ~/.bashrc
+conda activate credit
+torchrun applications/train.py -c model.yml
+```
+
+and note that the ```torchrun``` command is used rather than MPIs. For now MPIs are not supported on Casper but that may change in a future release. But to get torch to run in distributed mode (either DDP or FSDP) we use torchrun to faciliate that (rather than setting that up manuanlly in train.py). 
```

Comment:
Building PyTorch from source on casper would enable MPI support. torchrun works well-enough for single GPU applications.